src/Templates/AbstractTemplateCompatibility.php

```diff
@@ -0,0 +1,157 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Templates;
+
+/**
+ * BlockTemplatesCompatibility class.
+ *
+ * To bridge the gap on compatibility with PHP hooks and blockified templates.
+ *
+ * @internal
+ */
+abstract class AbstractTemplateCompatibility {
```

Comment:
I didn't know about trait keyword. I will check it

Reply:
Since this PR is already big, I plan to merge this when you approve and, after that, in another PR, rename `BlockTemplatesCompatibility` to `ArchiveProductTemplateCompatibility` and convert this class to a trait.

Reply:
After consideration, I think inheritance is still better option for this case. So marking this as resolved.

Reply:
> After consideration, I think inheritance is still better option for this case. So marking this as resolved.

Do you mind sharing more info about what made you change your mind here @dinhtungdu ?

Reply:
@nefeline Oh, I'm sorry I missed your [previous comment](https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1109777286). I changed my mind because:
- for both classes, we need to hook into the WordPress hook system (init method). We might want to add more classes, so duplicating the init method among classes is a code smell to me.
- IMO, inheritance is bad when it becomes a chain, in this case, we only have one level of inheritance, so inheritance becomes more attractive than `trait` to me. We should also mark classes that extend `AbstractTemplateCompatibility` as `final` to ensure that convention. What do you think?

Reply:
> @nefeline Oh, I'm sorry I missed your https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1109777286.

No worries & thanks for sharing more details!

> for both classes, we need to hook into the WordPress hook system (init method). We might want to add more classes, so duplicating the init method among classes is a code smell to me.

Agreed. On the other hand, we could also have a trait that doesn't include the init method.

> IMO, inheritance is bad when it becomes a chain. In this case, we only have one level of inheritance, so inheritance becomes more attractive than trait to me.

I'm also not a huge fan of inheritance for this same reason (when we end up with this chain of classes and have to chase down the parent). 

> We should also mark classes that extend AbstractTemplateCompatibility as final to ensure that convention.

My only "but" here would be: making sure every single new PR opened follows this convention and adds the `final` keyword can be easy to miss, whereas with a trait we wouldn't have to worry about that. Either way, I really don't have a strong preference in this case.